### ISSUE TITLE
Early game exploration

### ISSUE THREAD

**Issue by shindavid** [shindavid/AlphaZeroArcade#41]
TLDR; Implement dynamic move temperature like KataGo has. Experiment with different parameterizations, validate through experimentation.

---

My first AlphaZero run for c4 was slow to recognize the middle column as the best move, both on an empty board, and on a board where the first two moves were played in the middle. The below plot shows the progression of the policy evaluations on these board states over ~60 hours of training (>3.9M games, >55M positions), starting from a set of games generated by random play with perfect oracle labels.

![image](https://user-images.githubusercontent.com/5217927/215591587-a2b004ae-d7d6-4c73-9315-c92f034f852d.png)

Note that the later generations took significantly more epochs of training, of ever increasing batch sizes, than earlier generations  (it became harder and harder to pass the gating test). Using time as the x-axis likely would make the spike on the right look more gradual.

The perfect labels in the initial training data explain the good starting points on the left. We got unlucky, however, as the network quickly shifted to predict low policy outputs for the middle column in these board states. Such unluckiness should not be unexpected, particularly if we don't cheat by initializing our training data with perfect oracle-labeled data.

The real question is, **why was it so slow to recover?**

The main answer is likely that there was an insufficient set of forces that pushed the system to explore moves that the neural network deemed bad.

Specifically, I used a hard-coded move-temperature of 0.5. By contrast:

* AlphaZero used a temperature of 1.0 during the first 30 moves and epsilon thereafter
* KataGo gradually decayed its temperature from 0.8 to 0.2 throughout the game. Furthermore, it played the first `r` moves of the game directly proportional to the raw policy distribution, for a randomly chosen `r`.

Details can be found in Appendix D of the KataGo [paper](https://arxiv.org/pdf/1902.10565.pdf).

I chose to use a fixed move-temperature initially for simplicity, and picked 0.5 because it was in the middle of the temperature schedules of AlphaZero and KataGo.

I believe that using a smaller value than the starting temperatures of 1.0 and 0.8 for AlphaZero and KataGo, respectively, led to the network getting "stuck" believing in the superiority of non-middle openings, unwilling to sufficiently explore other options.

Another potential problem with the fixed 0.5 temperature is that later in the game, the agent may have been prone to _over_-explore, relative to the ending temperatures of epsilon and 0.2 for AlphaZero and KataGo, respectively. This could have led to noisier value labels, which could also contribute to the system being slower to recover.

Other factors like insufficient root-softmax-temperature and missing target-pruning likely contributed and are noted in Issues https://github.com/shindavid/blokus/issues/35 and https://github.com/shindavid/blokus/issues/24. But this move temperature issue I suspect is the biggest one, as AlphaZero was successful without those other mechanisms.

Furthermore, I wonder if the lower branching factor of c4 warrants different temperature choices to match an equivalent level of exploration as KataGo. The fact that David Wu chose **not** to incorporate the board width `b` into the temperature min/max values suggests to me that the answer is "no", as he appears to have been quite thoughtful about when to incorporate `b` into parameter schedules.
